packages/core/src/config/parse.ts

```diff
@@ -171,14 +198,507 @@ export const assertValidUserConfigFields = (config: UserChugSplashConfig) => {
+  // Hardhat's default `CompilerOutput`, which is what OpenZeppelin expects.
+  const dereferencer = astDereferencer(compilerOutput as any)
+  const extendedLayout = extendStorageLayout(storageLayout, dereferencer)
+
+  for (const variableName of Object.keys(contractConfig.variables)) {
+    const existsInLayout = extendedLayout.storage.some(
+      (storageObj) => storageObj.configVarName === variableName
+    )
+
+    if (existsInLayout === false) {
+      // Complain very loudly if attempting to set a variable that doesn't exist within this layout.
+      throw new Error(
+        `Variable "${variableName}" was defined in the ChugSplash config file for ${contractConfig.contract} but\n` +
+          `does not exist as a mutable variable in the contract. If "${variableName}" is immutable, please remove\n` +
+          `its definition in the 'variables' section of the ChugSplash config file and use the 'constructorArgs' field\n` +
+          `instead. If this variable is not meant to be immutable, please remove this variable definition in the\n` +
+          `ChugSplash config file. If this problem persists, delete your cache folder then try again.`
+      )
+    }
+  }
+
+  for (const storageObj of Object.values(extendedLayout.storage)) {
+    const configVarValue = contractConfig.variables[storageObj.configVarName]
+    if (configVarValue === undefined) {
+      throw new Error(
+        `Detected a variable "${storageObj.configVarName}" from the contract "${contractConfig.contract}" (or one\n` +
+          `of its parent contracts), but could not find a corresponding variable definition in your ChugSplash config.\n` +
+          `file. Every variable defined in your contracts must be assigned a value in your ChugSplash config file.\n` +
+          `Please define the variable in your ChugSplash config file then run this command again.\n` +
+          `If this problem persists, delete your cache folder then try again.`
+      )
+    }
+
+    try {
+      parsedConfigVariables[storageObj.configVarName] =
+        parseAndValidateVariable(
+          configVarValue,
+          storageObj,
+          extendedLayout.types,
+          '0',
+          dereferencer
+        )
+    } catch (e) {
+      errors.push((e as Error).message)
+    }
+  }
+
+  if (errors.length > 0) {
+    let message = `We detected some issues in your ChugSplash config file, please resolve them and try again.\n\n`
+    for (const error of errors) {
+      message += `${error} \n`
+    }
+    throw new Error(message)
+  }
+
+  return parsedConfigVariables
+}
+
 /**
  * Parses a ChugSplash config file from the config file given by the user.
  *
  * @param userConfig Unparsed config file to parse.
  * @param env Environment variables to inject into the file.
  * @return Parsed config file with template variables replaced.
  */
-export const parseChugSplashConfig = async (
+const parseAndValidateChugSplashConfig = async (
```

Comment:
add `export`?

Reply:
This function is not supposed to be used outside this library. Instead, we should use `readParsedChugSplashConfig` which includes more validation and calls `parseAndValidateChugSplashConfig`.